gr-blocks/lib/correctiq_man_impl.cc

```diff
@@ -0,0 +1,127 @@
+#include <gnuradio/io_signature.h>
+
+namespace gr {
+namespace blocks {
+
+correctiq_man::sptr correctiq_man::make(float real, float imag)
+{
+    return gnuradio::get_initial_sptr(new correctiq_man_impl(real, imag));
+}
+
+/*
+ * The private constructor
+ */
+correctiq_man_impl::correctiq_man_impl(float real, float imag)
+    : gr::sync_block("correctiq_man",
+                     gr::io_signature::make(1, 1, sizeof(gr_complex)),
+                     gr::io_signature::make(1, 1, sizeof(gr_complex)))
+{
+    avg_real = real;
+    avg_img = imag;
+    d_k = gr_complex(avg_real, avg_img);
+
+    message_port_register_in(pmt::mp("set_real"));
+    set_msg_handler(pmt::mp("set_real"),
+                    [this](pmt::pmt_t msg) { this->handle_real(msg); });
+    message_port_register_in(pmt::mp("set_imag"));
+    set_msg_handler(pmt::mp("set_imag"),
+                    [this](pmt::pmt_t msg) { this->handle_imag(msg); });
+}
+
+/*
+ * Our virtual destructor.
+ */
+correctiq_man_impl::~correctiq_man_impl() {}
+
+float correctiq_man_impl::get_real() { return avg_real; }
+float correctiq_man_impl::get_imag() { return avg_img; }
+
+void correctiq_man_impl::set_real(float new_value)
+{
+    avg_real = new_value;
+    d_k = gr_complex(avg_real, avg_img);
+}
+
+void correctiq_man_impl::set_imag(float new_value)
+{
+    avg_img = new_value;
+    d_k = gr_complex(avg_real, avg_img);
+}
+
+void correctiq_man_impl::handle_real(pmt::pmt_t msg)
+{
+    if (pmt::is_pair(msg)) {
+        pmt::pmt_t data = pmt::cdr(msg);
+        if (pmt::is_real(data)) {
+            float new_value = pmt::to_float(data);
+
+            set_real(new_value);
+        } else {
+            GR_LOG_WARN(d_logger, "Non-float real value received.  Ignoring.");
+        }
+    } else {
+        if (pmt::is_real(msg)) {
+            float new_value = pmt::to_float(msg);
+
+            set_real(new_value);
+        } else {
+            GR_LOG_WARN(d_logger, "Non-float real value received.  Ignoring.");
+        }
+    }
+}
+
+void correctiq_man_impl::handle_imag(pmt::pmt_t msg)
+{
+    if (pmt::is_pair(msg)) {
+        pmt::pmt_t data = pmt::cdr(msg);
+        if (pmt::is_real(data)) {
+            float new_value = pmt::to_float(data);
+
+            set_imag(new_value);
+        } else {
+            GR_LOG_WARN(d_logger, "Non-float imag value received.  Ignoring.");
+        }
+    } else {
+        if (pmt::is_real(msg)) {
+            float new_value = pmt::to_float(msg);
+
+            set_imag(new_value);
+        } else {
+            GR_LOG_WARN(d_logger, "Non-float imag value received.  Ignoring.");
+        }
+    }
+}
+
+int correctiq_man_impl::work(int noutput_items,
+                             gr_vector_const_void_star& input_items,
+                             gr_vector_void_star& output_items)
+{
+    const gr_complex* in = (const gr_complex*)input_items[0];
+    gr_complex* out = (gr_complex*)output_items[0];
+
+    for (int i = 0; i < noutput_items; i++) {
```

Comment:
Guessing this would be faster using Volk.

Reply:
Do you want to Volk-ify this? I'm happy to run you through the steps ... this one is easy-peasy & would be a lot faster!